mining: getCoinbase() returns struct instead of raw tx

Sjors opened the pull request:
The first commit renames `getCoinbaseTx()` to `getCoinbaseRawTx()` to reflect that it returns a serialised transaction. This does not impact IPC clients, because they do not use the function name.

The second commit then introduces a replacement `getCoinbase()` that provides a struct with everything clients need to construct a coinbase. This avoids clients having to parse and manipulate our dummy transaction.

Deprecate but don't remove `getCoinbaseRawTx()`, `getCoinbaseCommitment()` and `getWitnessCommitmentIndex()`.

After this change we can drop these deprecated methods, which in turn would allow us to clear the dummy transaction from the `getBlock()` result. But that is left for a followup to keep this PR focussed. See https://github.com/Sjors/bitcoin/pull/106 for an approach.

Expand the `interface_ipc.py` functional test to document its usage.

Can be tested using:
- https://github.com/stratum-mining/sv2-tp/pull/59

Sjors commented:
Working on a matching [sv2-tp](https://github.com/stratum-mining/sv2-tp) pull request. It will first try `getCoinbase()` and if that fails it falls back to `getCoinbaseTx()`.

ismaelsadeeq commented on src/interfaces/mining.h:
```diff
@@ -42,8 +42,26 @@ class BlockTemplate
     // Sigop cost per transaction, not including coinbase transaction.
     virtual std::vector<int64_t> getTxSigops() = 0;
 
+    /** Return fields needed to construct a coinbase transaction */
```
Given [sv2-tp](https://github.com/stratum-mining/sv2-tp) and a few others are the users of this interface and they are still in development.

I'd rather have any un-useful methods pruned.
When we release a stable interface then we can start deprecating things to avoid breaking clients (As you mentioned in the tracking issue it is okay to make breaking changes so all these methods should be pruned and clients should update).

ismaelsadeeq replied:
Hmm, do you plan on deprecating it first and then removing it in a later release? On the contrary, I think it should be dropped as soon as we deem it unnecessary. Otherwise, we’ll end up with interface methods that conflict with each other. It would also be easier to review the changes together, rather than keeping them all here with deprecation comments and later deleting them it will save us review cycle.

> Even for testing it's nice that current clients works against both v30 and master.

I think this is not big of deal, since you have a release that is against v30, your master should be  in sync with master which can easily update when breaking change occur.

Sjors replied:
> do you plan on deprecating it first and then removing it in a later release?

No, because the interface is marked experimental I think it's fine to drop it in any release we want. That might be as soon as v31, but we can also do it later.

> your master should be in sync with master

`sv2-tp` is used by people who only run v30, so it has to support both master and v30.

It can already handle _new_ methods by just trying them and falling back to old methods when they're missing. 

Recent versions of `sv2-tp` all work against v30, so I haven't had to push people to switch to a newer version either. But we start making breaking changes, then I have to start instructing users of Bitcoin Core master to use specific newer versions of `sv2-tp`.

ismaelsadeeq replied:
> In the longer run it probably makes sense for sv2-tp to have one major release of sv2-tp for every major Bitcoin Core release, but current it doesn't have backports. Each release is a combination of bug fixes and new features.


Hmm, fwiw I am approach NACK changes that introduce conflicting interface method that deprecate and don't prune on experimental interface that did not offer interface stability for the reason discussed above.
However I am open minded to convincing otherwise and curious to know what other contributors think.

Sjors replied:
The old methods are redundant, but they're not conflicting. Clients can use either and both have test coverage.

From this projects perspective there's hardly any additional review burden when we prune deprecated methods in a later PR, that should be trivial.

Sjors commented:
@plebhash can you try to see if it's easy for you to support both `getCoinbase()` and `getCoinbaseTx()` in https://github.com/stratum-mining/sv2-apps/pull/59, preferring the first and falling back to the latter if it doesn't exist?

optout21 commented on test/functional/interface_ipc.py:
```diff
@@ -188,10 +198,56 @@ async def async_routine():
-            coinbase = await self.parse_and_deserialize_coinbase_tx(template, ctx)
-            balance = miniwallet.get_balance()
+
+            self.log.debug("Build coinbase transaction using getCoinbase()")
```
Maybe extract this part into a separate method; that can also serve as an example on how to build the cb tx from the getCoinbase() result, and leave the flow of this test more readable

optout21 commented on src/node/interfaces.cpp:
As a progression towards `getCoinbase` & deprecation, this implementation could invoke ExtractCoinbaseTemplate`, and return the `witness` field (after uint256 -> unsigned char array type conversion)

Sjors replied:
Refactoring these deprecated methods would add to the review burden. And there's still ongoing discussion above about whether this PR should even keep them

optout21 commented on src/interfaces/mining.h:
```diff
@@ -42,8 +42,26 @@ class BlockTemplate
+    /**
+     * Return scriptPubKey with SegWit OP_RETURN.
+     *
+     * @note deprecated: use the output field from getCoinbase()
```
The `witness` field could also be used, isn't it?

optout21 commented on src/node/interfaces.cpp:
As a progression towards `getCoinbase` & deprecation, this implementation could internally also do what is suggested in the deprecation node:
invoke `ExtractCoinbaseTemplate`, scan outputs for the SegWit marker, and return the index.

Sjors commented:
Odd, the invalid-version error is happening at the bottom of this block:

```py
self.log.debug("Submit a block with a bad version")
block.nVersion = 0
block.solve()
res = await mining.result.checkBlock(block.serialize(), check_opts)
assert_equal(res.result, False)
assert_equal(res.reason, "bad-version(0x00000000)")
res = await template.result.submitSolution(ctx, block.nVersion, block.nTime, block.nNonce, coinbase.serialize())
assert_equal(res.result, False)
self.log.debug("Submit a valid block")
block.nVersion = original_version
block.solve()
res = await mining.result.checkBlock(block.serialize(), check_opts)
assert_equal(res.result, True)
```

It's about the block `nVersion`, not the coinbase, so this PR shouldn't have changed that.

Sjors commented:
It turns out that using the `getCoinbase()` result directly in the Python test led to memory management related errors. Parts of the coinbase were dropped by the time of the second `checkBlock()`.

Rather than debug (our use of) PyCapnp, I just introduced a `@dataclass` for `CoinbaseTemplateData`, so we fully own it.

Sjors commented:
> we do getBlock and that provides everything we need

Ok, that's not really how you're supposed to use it :-)

It does make a breaking change potentially unpainful.

However there's another gotcha if we remove a method from the interface: there's no gaps allowed in the `.capnp` method numbering. This could be worked around as follows:

```capnp
interface BlockTemplate $Proxy.wrap("interfaces::BlockTemplate") {
    destroy @0 (context :Proxy.Context) -> ();
    getBlockHeader @1 (context: Proxy.Context) -> (result: Data);
    getBlock @2 (context: Proxy.Context) -> (result: Data);
    getTxFees @3 (context: Proxy.Context) -> (result: List(Int64));
    getTxSigops @4 (context: Proxy.Context) -> (result: List(Int64));
    getCoinbase @5 (context: Proxy.Context) -> (result: CoinbaseTemplate);
    # getCoinbaseTx @5 (context: Proxy.Context) -> (result: Data);
    # getCoinbaseCommitment @6 (context: Proxy.Context) -> (result: Data);
    getWitnessCommitmentIndex @7 (context: Proxy.Context) -> (result: Int32);
    getCoinbaseMerklePath @8 (context: Proxy.Context) -> (result: List(Data));
    submitSolution @9 (context: Proxy.Context, version: UInt32, timestamp: UInt32, nonce: UInt32, coinbase :Data) -> (result: Bool);
    waitNext @10 (context: Proxy.Context, options: BlockWaitOptions) -> (result: BlockTemplate);
    interruptWait @6() -> ();
}
```

So `@5` is assigned from `getCoinbaseTx` to `getCoinbase`, `@6` is changed from `getCoinbaseCommitment` to `interruptWait` (not in v30 and not yet in an sv2-tp release), but `getWitnessCommitmentIndex` is kept (but still deprecated and made to error).

But such an intermediate state seems needlessly confusing, so my first preference is still to not make a breaking change here, my second preference to completely break things, not this intermediate approach.

ryanofsky commented on src/node/miner.cpp:
```diff
@@ -586,4 +586,43 @@ std::optional<BlockRef> WaitTipChanged(ChainstateManager& chainman, KernelNotifi
+    // The CoinbaseTemplate interface guarantees a size limit. Raising it (e.g.
+    // if a future softfork needs to commit more than BIP34) is a
+    // breaking change for clients.
+    Assume(coinbase.script_sig_prefix.size() <= 8);
```
In commit "mining: add getCoinbase()" (80acabaf9353245f0839b18d9552ab39f595eda6)

It seems like if this change could break clients, it might be better to return an error than a template that violates their expectations. Maybe would suggest:

```c++
if (!Assume(coinbase.script_sig_prefix.size() <= 8)) {
  throw std::logic_error("Internal bug: coinbase data is inconsistent with CoinbaseTemplate definition. Data may be corrupt or definition may need to be updated");
}
```

Or could replace Assume with Assert, or log a warning. it just doesn't seem good to ignore the condition completely if it happens.

ryanofsky commented on src/node/miner.h:
```diff
@@ -259,6 +259,17 @@ std::optional<BlockRef> GetTip(ChainstateManager& chainman);
+ * Extracts only OP_RETURN coinbase outputs, i.e. the witness commitment. If
+ * BlockAssembler::CreateNewBlock() is patched to add more OP_RETURN outputs
+ * e.g. for merge mining, those will be included. The dummy output that spends
+ * the full reward is excluded.
```
In commit "mining: add getCoinbase()" (80acabaf9353245f0839b18d9552ab39f595eda6)

I think this comment might would be more helpful if moved to the `CoinbaseTemplate::output` field since it describes what that field contains. Also it should be visible there since that type is part of the external interface.

Sjors replied:
I added documentation there, but we also need it here. The `output` field is not limited to `OP_RETURN`, that's just a heuristic used by `ExtractCoinbaseTemplate`.

That heuristic won't work anymore if we add a way for clients to specify custom outputs when constructing a block, so it's important to document.

ryanofsky commented on src/node/types.h:
```diff
@@ -99,6 +102,35 @@ struct BlockCheckOptions {
+     *
+     * This is currently the BIP 141 witness reserved value. A future soft fork
+     * may move the witness reserved value elsewhere, but there will still be a
+     * coinbase witness.
```
In commit "mining: add getCoinbase()" (80acabaf9353245f0839b18d9552ab39f595eda6)

I'm a little confused by "may move the witness reserved value" in this comment. I would think a future soft fork could restrict this value but not move it anywhere. BIP 141 says the coinbase witness has to be a single 32-byte so that doesn't seem to allow any other flexibility.

Sjors replied:
The BIP says:

> For backward compatibility, the `Hash(new commitment|witness reserved value)` will go to the coinbase witness, and the witness reserved value will be recorded in another location specified by the future softfork. Any number of new commitment could be added in this way.

So we provide clients with the coinbase witness, which is not necessarily the "witness reserved value".

ryanofsky commented on src/node/types.h:
```diff
@@ -99,6 +102,35 @@ struct BlockCheckOptions {
+    /* nVersion */
+    uint32_t version;
+    /* nSequence for the only coinbase transaction input */
+    uint32_t input_sequence;
```
In commit "mining: add getCoinbase()" (80acabaf9353245f0839b18d9552ab39f595eda6)

Seems like it would be more consistent to either drop the `input_` prefix from this field or add it to `script_sig_prefix` field. I'd slightly prefer dropping the `input_` field so `CoinbaseTemplate` field names describe what they mean, not how the transaction needs to be physically constructed.

ryanofsky commented on src/node/types.h:
```diff
@@ -99,6 +102,35 @@ struct BlockCheckOptions {
+     * may move the witness reserved value elsewhere, but there will still be a
+     * coinbase witness.
+     */
+    std::optional<uint256> witness;
```
In commit "mining: add getCoinbase()" (80acabaf9353245f0839b18d9552ab39f595eda6)

IMO `witness_reserved_value` would be a more descriptive name than just `witness`.

ryanofsky reviewed (approved):
Code review ACK 80acabaf9353245f0839b18d9552ab39f595eda6. I left various suggestions below but none are critical. This seems like a thoughtful change that should make clients easier to write and be compatible with potential soft forks, without placing a lot of unnecessary requirements on them.

---

re: https://github.com/bitcoin/bitcoin/pull/33819#issuecomment-3520618909

> However there's another gotcha if we remove a method from the interface: there's no gaps allowed in the .capnp method numbering. This could be worked around as follows

Agree with you that workaround seems pretty confusing and would not recommend it.

Also agree with you in https://github.com/bitcoin/bitcoin/pull/33819#discussion_r2504862983 that it makes sense to deprecate rather than remove the old methods for the time being.

When it actually comes time to remove the methods I think there are three potential approaches:

1 - Bump the Init.makeMining() ordinal and return an error to old clients as soon as they connect. This way we are free to renumber everything in the Mining interface and make other backwards incompatible changes, because old clients will never gain access to the new interface.
2 - Define some annotation like `$removed` that could return "method not implemented" or errors to clients that attempt to call them.
3 - Just renumber the interface without caring about compatibility with old clients.

I think approach (1) makes sense if there are other backwards incompatible changes we want to make and batch together. Otherwise approach (2) is probably preferable so clients like plebhash's that do not call the removed methods are not affected. Approach (3) is probably best to avoid, but it is an option. I also think it would be fine to keep the deprecated methods around indefinitely, since they are very simple, as long as they are renamed to discourage use in new code.

Sjors commented:
From https://github.com/bitcoin/bitcoin/issues/33777#issuecomment-3528997008 I gather that we should not reuse the method ordinals. Since cap&proto doesn't allow gaps, when dropping a deprecated method, it's best to just have them throw an error.

In https://github.com/bitcoin/bitcoin/issues/33777#issuecomment-3529071835 @ryanofsky suggested replacing the dummy coinbase with an empty transaction, so that clients are not tempted to use the dummy coinbase via `getBlock()` or `getCoinbaseTx()`.

I think I'm going to combine these two observations and make this PR a breaking change. Marking as draft for now.

ryanofsky commented:
> I think I'm going to combine these two observations and make this PR a breaking change. Marking as draft for now.

This seems ok, but isn't really what I would suggest. My suggestion would be for this PR to just introduce a new `getCoinBase()` method like it's doing and not remove other methods so existing mining clients do not immediately stop working. The other methods, which are all simple accessors, can easily be removed later.

Adding a new method now and removing older methods later should make both PR's simpler and more focused. It would also let existing clients continue to work with the latest version of bitcoin while new clients are developed. The second PR removing the deprecated methods could also swap the coinbase TX with an empty one as suggested to avoid getBlock returning it.

The only additional thing I might consider doing in this PR (in a second commit) is renaming methods intended to be deprecated, as shown in the diff https://github.com/bitcoin/bitcoin/pull/33819#discussion_r2519222031. But even this could be overkill. It should be fine to just add `getCoinBase()` here and remove other ways of accessing coinbase info in a followup. Ok to take a different approach though. Just wanted to clarify my opinion.

Sjors commented:
> The second PR removing the deprecated methods could also swap the coinbase TX with an empty one as suggested to avoid getBlock returning it.

It makes sense to split this idea over two PR's. I have a local branch already, but still need to split out the commits and get the commit order right.

Sjors commented:
@ajtowns the `CoinbaseTemplate` struct introduced by this PR closely matches the [NewTemplate message](https://stratumprotocol.org/specification/07-Template-Distribution-Protocol/#72-newtemplate-server-client) in Stratum v2. The main difference is that the `coinbase_tx_outputs` field of `NewTemplate` uses an ad hoc serialisation that I don't think we should embrace, so I keep the outputs in a vector.

Note that the code in `ExtractCoinbaseTemplate()` was mostly copied from `sv2-tp`. So the main difference is that we now control how our dummy coinbase is "mined" for fields, rather that hope that clients implement it correctly.

> Constructing a coinbase requires getting various consensus requirements correct, and new requirements may be added in future, such as the nlocktime requirement proposed in bip54.

Unless we overhaul the sv2 [NewTemplate message](https://stratumprotocol.org/specification/07-Template-Distribution-Protocol/#72-newtemplate-server-client) it doesn't help if our code implements such a new rule correctly, because the IPC client (like sv2-tp) might ignore the new field when it takes apart the dummy coinbase.

In the short run, bip54 is already taken care of because `coinbase_tx_locktime` is a field (whereas `getblocktemplate` doesn't have this yet).

If a future soft fork adds a new consensus required field, we would have to add it to the `CoinbaseTemplate` struct (not a breaking change), inform Stratum v2 protocol designers that they need to add it to `NewTemplate` and any downstream tooling.

However, I can't think of anything that we can add to the coinbase that's not covered by existing fields:

- version is provided
- a coinbase can only have one input
  - its `prevhash` and `n` are pinned by consensus
  - witness must be a single 32 byte push by consensus, we provide the value
  - `scriptSig` (coinbase) prefix is provided (making it longer than 8 bytes is a breaking change in sv2)
  - sequence is provided
- all mandatory outputs are provided
- lockTime is provided

I like the idea of replacing `m_options.coinbase_output_script` with a vector. It would allow the client to add merge mining outputs and potentially avoid patching Bitcoin Core.

Sjors commented:
@plebhash wrote:

> however, we always need to call getBlock specifically because it contains the header + txdata of the template... and as a byproduct, we happen to also get the Coinbase Tx data, which saves us one extra call to getCoinbase/getCoinbaseTx

Regarding the header, there's a separate `getHeader()` method.

It's true that you need txdata anyway, but you don't need it immediately. You need it when the pool asks using the `RequestTransactionData` message.

When the node and the Template Provider run on the same machine, none of this matters. IPC round trips are very fast and it doesn't matter if you fetch the whole block via `getBlock()` and then extract the header, or if you first do `getHeader()` and do `getBlock()` later.

The interface design is mostly based on the Stratum v2 message design and that design in turn is built on the assumption that network latency _does_ matter. So you want to be able to very quickly construct and broadcast the `NewTemplate` message, while taking your time with `RequestTransactionData.Success` and getting a template approved by the pool.

> but it will take a while before we start exploring this avenue, so for now it's probably better for us to stick with getBlock (which I assume will remain stable?

This pull request doesn't break `getBlock()`, but it's possible that v31 will break it by wiping the dummy coinbase (in a followup PR).

Sjors commented on src/node/miner.cpp:
```diff
@@ -586,4 +586,43 @@ std::optional<BlockRef> WaitTipChanged(ChainstateManager& chainman, KernelNotifi
+
+    // The first dummy coinbase output produced by CreateBlock() has an nValue
+    // set to nFee + the Block Subsidy
+    coinbase.value_remaining = coinbase_tx->vout[0].nValue;
```
I'll go for the latter approach.

Sjors commented on src/node/types.h:
```diff
@@ -99,6 +102,35 @@ struct BlockCheckOptions {
+     * coinbase witness.
+     */
+    std::optional<uint256> witness;
+    // Block subsidy plus fees, minus any non-zero coinbase outputs
```
Changed to "minus any non-zero required outputs" and a comment to point out that those don't exist atm.

See https://github.com/bitcoin/bitcoin/pull/33819#discussion_r2533732177 above, I think having non-zero outputs is quite conceivable and it would be a shame if that requires another breaking change.

Sjors commented:
> OTOH, overhauling NewTemplate shouldn't be a blocker for supporting sv2 efficiently, so Approach ACK for this PR in the meantime.

Thanks, I opened https://github.com/stratum-mining/sv2-spec/issues/167 to continue that discussion.

>> replacing m_options.coinbase_output_script with a vector.

See https://github.com/bitcoin/bitcoin/pull/33890

Sjors commented on src/node/miner.cpp:
```diff
@@ -591,4 +591,49 @@ std::optional<BlockRef> WaitTipChanged(ChainstateManager& chainman, KernelNotifi
+    // mining, etc). BlockAssembler::CreateNewBlock adds a dummy output with
+    // the full reward that we must exclude.
+    for (const auto& output : coinbase_tx->vout) {
+        if (!output.scriptPubKey.empty() && output.scriptPubKey[0] == OP_RETURN) {
```
Done in the latest push.

ryanofsky commented on src/node/miner.cpp:
```diff
@@ -591,4 +594,44 @@ std::optional<BlockRef> WaitTipChanged(ChainstateManager& chainman, KernelNotifi
+    if (!Assume(coinbase.script_sig_prefix.size() <= 8)) {
+        LogWarning("Unexpected %d byte scriptSig prefix size.",
+                    coinbase.script_sig_prefix.size());
+    }
```
Instead of having a warning here it would seem better to document mining client reliance on scriptSig length directly in the code generating the scriptsig. This way, if there is a problematic change it can be obvious from the code, and not rely on noticing a warning. Also it would seem clearer and more future proof for IPC documentation to just say that the final scriptsig needs to be <= 100 bytes, instead of trying to "guarantee" miners can always add 92 bytes.

So would suggest a change like:

```diff
--- a/src/node/miner.cpp
+++ b/src/node/miner.cpp
@@ -168,6 +168,11 @@ std::unique_ptr<CBlockTemplate> BlockAssembler::CreateNewBlock()
     coinbaseTx.vout.resize(1);
     coinbaseTx.vout[0].scriptPubKey = m_options.coinbase_output_script;
     coinbaseTx.vout[0].nValue = nFees + GetBlockSubsidy(nHeight, chainparams.GetConsensus());
+    // Start the coinbase scriptSig with the block height as required by BIP34.
+    // The trailing OP_0 is optional padding and could be removed without a
+    // consensus change. Mining clients are expected to append extra data to
+    // this prefix, so increasing its length would reduce the space they can use
+    // and may break existing clients.
     coinbaseTx.vin[0].scriptSig = CScript() << nHeight << OP_0;
     Assert(nHeight > 0);
     coinbaseTx.nLockTime = static_cast<uint32_t>(nHeight - 1);
@@ -603,13 +608,6 @@ node::CoinbaseTemplate ExtractCoinbaseTemplate(const node::CBlockTemplate& block
     coinbase.version = coinbase_tx->version;
     Assert(coinbase_tx->vin.size() == 1);
     coinbase.script_sig_prefix = coinbase_tx->vin[0].scriptSig;
-    // The CoinbaseTemplate interface guarantees a size limit. Raising it (e.g.
-    // if a future softfork needs to commit more than BIP34) is a
-    // (potentially silent) breaking change for clients.
-    if (!Assume(coinbase.script_sig_prefix.size() <= 8)) {
-        LogWarning("Unexpected %d byte scriptSig prefix size.",
-                    coinbase.script_sig_prefix.size());
-    }
 
     if (coinbase_tx->HasWitness()) {
         const auto& witness_stack{coinbase_tx->vin[0].scriptWitness.stack};
--- a/src/node/types.h
+++ b/src/node/types.h
@@ -108,9 +108,11 @@ struct CoinbaseTemplate {
     /* nSequence for the only coinbase transaction input */
     uint32_t sequence;
     /**
-     * Bytes which are to be placed at the beginning of scriptSig. Guaranteed
-     * to be less than 8 bytes (not including the length byte). This allows
-     * clients to add up to 92 bytes.
+     * Prefix which needs to be placed at the beginning of the scriptSig.
+     * Clients may append extra data to this as long as the overall scriptSig
+     * size is 100 bytes or less, to avoid the block being rejected with
+     * "bad-cb-length" error. Currently with BIP 34, the prefix should be less
+     * than 8 bytes, but future changes could require longer prefixes.
      */
     CScript script_sig_prefix;
     /**
```
<details><summary>diff</summary>
<p>

```diff

```
</p>
</details>

ryanofsky reviewed (approved):
Code review ACK 73553c809ef6517839c4d1b02b657e6fa401f8a8. I left some minor suggestions below but they are not important. Thanks for responding to all the previous ones.

AJ's bigger suggestions https://github.com/bitcoin/bitcoin/pull/33819#issuecomment-3539952304 and https://github.com/bitcoin/bitcoin/pull/33819#issuecomment-3545336207 definitely make a lot of sense to me, and would seem be a simpler & more future-proof direction to go in. They seem directionally compatible with this change, though. They could be implemented by expanding the `BlockCreateOptions` struct and changing the `CoinbaseTemplate` struct to have `prefix` and `suffix` fields instead of `version`, `sequence`, etc fields, I think.

I do have one question about the implementation of this PR, not the interface: It seems like the `ExtractCoinbaseTemplate` function would be unnecessary if you would add the 
`CoinbaseTemplate m_coinbase_template` field to the `CBlockTemplate` class instead of the `BlockTemplateImpl` class. Then the `BlockAssembler::CreateNewBlock` method could just fill the coinbase template out directly and not require another method to extract it later. Wonder if you considered this or think it might be a good idea.

Sjors committed:
mining: rename getCoinbaseTx() to ..RawTx()

This frees up the name getCoinbaseTx() for the next commit.

Changing a function name does not impact IPC clients, as they only
consider the function signature and sequence number.

ryanofsky commented on src/node/miner.cpp:
```diff
@@ -158,18 +158,52 @@ std::unique_ptr<CBlockTemplate> BlockAssembler::CreateNewBlock()
+        coinbase_tx_template.required_outputs.push_back(final_coinbase->vout[witness_index]);
+    }
+
+    pblocktemplate->m_coinbase_tx_template = std::move(coinbase_tx_template);
```
In commit "mining: add new getCoinbaseTx() returning a struct" (60f74dd315811db427797a606f4b5611a8c59993)

Would be nice to avoid the move and treat `pblock` and `coinbase_tx_template` local variables more consistently. Would suggest defining a reference like `CoinbaseTxTemplate& coinbase_tx_template{pblocktemplate->m_coinbase_tx_template}` above similar to `pblock`.

ryanofsky reviewed (approved):
Code review ACK 60f74dd315811db427797a606f4b5611a8c59993. New comments and renaming both seem great, and it is nice to get rid of the extract function. PR feels much more straightforward now.

Sjors commented:
Applied both suggestions from @ryanofsky: https://github.com/bitcoin/bitcoin/pull/33819#pullrequestreview-3536150722

ryanofsky reviewed (approved):
Code review ACK 5761dac2f5481c27550483279a007c2b850bc990, with the two suggestions applied since last review.

ryanofsky reviewed (approved):
Code review ACK 1bd6cae51fedeff62755e03ddcb9346b868f6d83. Just rebased with no changes, to avoid trivial merge conflict.

ryanofsky reviewed (approved):
Code review ACK 1bd6cae51fedeff62755e03ddcb9346b868f6d83. Just rebasing again since last review to avoid minor conflict

Sjors commented:
Rebased after https://github.com/bitcoin/bitcoin/pull/34003.

ismaelsadeeq commented on src/ipc/capnp/mining.capnp:
```diff
@@ -51,3 +52,13 @@ struct BlockCheckOptions $Proxy.wrap("node::BlockCheckOptions") {
+    sequence @1 :UInt32 $Proxy.name("sequence");
+    scriptSigPrefix @2 :Data $Proxy.name("script_sig_prefix");
+    witness @3 :Data $Proxy.name("witness");
+    blockRewardRemaining @4 :Int64 $Proxy.name("block_reward_remaining");
```
In 979805b81c81db4c83cf5d9ed716a5c88bf469b0 "mining: add new getCoinbaseTx() returning a struct"

Why is this named blocks remaining?

Sjors replied:
See this thread: https://github.com/bitcoin/bitcoin/pull/33819#discussion_r2519805319

ismaelsadeeq commented on src/node/types.h:
```diff
@@ -99,6 +102,48 @@ struct BlockCheckOptions {
+     * softforks or a custom mining patch could add more.
+     *
+     * The dummy output that spends the full reward is excluded.
+     */
```
In 979805b "mining: add new getCoinbaseTx() returning a struct"

nit: Current comment is a bit speculative. I think we should just document existing rules and requirements.

Sjors replied:
The reason it's a bit more elaborate is that otherwise people will be confused why this is an array.

The part about merged mining is not speculative, only the part about future soft-forks. But from the point of view of a client implementer this speculation is very important to keep in mind. They should not be lazy and only take the first element from the array. And by "they" I mean myself :-) https://github.com/stratum-mining/sv2-tp/pull/55

ismaelsadeeq commented on src/node/miner.h:
```diff
@@ -50,6 +50,11 @@ struct CBlockTemplate
+    /*
+     * Coinbase transaction template, so IPC clients don't have to parse the
+     * dummy transaction directly.
+     */
```
In 979805b "mining: add new getCoinbaseTx() returning a struct"


```suggestion
/*
     * Coinbase transaction template includes all the
     * fields necessary to construct a consensus valid transaction.
     */
```

Sjors replied:
Went for a slight variation since not all values are mandatory. More accurate is to say that they're set by our mining code.

ismaelsadeeq reviewed (commented):
ACK 979805b81c81db4c83cf5d9ed716a5c88bf469b0

The commit message is helpful for my review. 

I still prefer that deprecated methods be removed once we have a replacement for them, for 1. better hygiene for the codebase. 2 Ensures we only maintain what should be used and our perceived valid approach for IPC clients. 3. Will prevent naive users from using deprecated method.

However, others seem to prefer deprecating them first, then batch-removing them later.

https://github.com/bitcoin/bitcoin/pull/33819#pullrequestreview-3454679064 Option 2 is my second preferred option. The method names and numbering are kept, but we return an error when someone tries to use them
It would allow us to remove the implementation code and tests while maintaining the methods and numbering for the transition period (can even be kept indefinitely since no implementation and test to maintain, could also be deleted in batch).

Sjors committed:
mining: add new getCoinbaseTx() returning a struct

Introduce a new method intended to replace getCoinbaseRawTx(), which
provides a struct with everything clients need to construct a coinbase.
This is safer than providing a raw dummy coinbase that clients then have
to manipulate.

The CoinbaseTx data is populated during the dummy transaction generation
and stored in struct CBlockTemplate.

Expand the interface_ipc.py functional test to document its usage
and ensure equivalence.

ryanofsky reviewed (approved):
Code review ACK 48f57bb35bbdbce509b8ef195de69e2a61a2511e. Just rebased and addressed comments and dropped coinbase tx "template" suffix, which is a nice change

ismaelsadeeq reviewed (approved):
code review ACK 48f57bb35bbdbce509b8ef195de69e2a61a2511e

vasild reviewed (approved):
ACK 48f57bb35bbdbce509b8ef195de69e2a61a2511e
